Integrate installation instructions into documentation (#6814)
cvc5/cvc5: INSTALL.rst

The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
library can be found in ``<build_dir>/lib``.

Build dependencies
------------------

cvc5 makes uses of a number of tools and libraries. Some of these are required
while others are only used with certain configuration options. If
``--auto-download`` is given, cvc5 can automatically download and build most
libraries that are not already installed on your system. If your libraries are
installed in a non-standard location, you can use ``--dep-path`` to define an
additional search path for all dependencies. Versions given are minimum
versions; more recent versions should be compatible.

- `GNU C and C++ (gcc and g++) <https://gcc.gnu.org>`_
  or `Clang <https://clang.llvm.org>`_ (reasonably recent versions)
- `CMake >= 3.9 <https://cmake.org>`_
- `Python 3.x <https://www.python.org>`_
  + module `toml <https://pypi.org/project/toml/>`_
- `GMP v6.1 (GNU Multi-Precision arithmetic library) <https://gmplib.org>`_
- `ANTLR 3.4 <http://www.antlr3.org/>`_
- `Java >= 1.6 <https://www.java.com>`_
- `SymFPU <https://github.com/martin-cs/symfpu/tree/CVC4>`_

> 

Reviewer: CaDiCaL is now also a required dependency.

ANTLR 3.4 parser generator
^^^^^^^^^^^^^^^^^^^^^^^^^^

For most systems, the package manager no longer contains pre-packaged versions
of ANTLR 3.4. With ``--auto-download``, cvc5 will automatically download and
build ANTLR 3.4.

GMP (GNU Multi-Precision arithmetic library)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

GMP is usually available on your distribution and should be used from there. If
it is not, or you want to cross-compile, or you want to build cvc5 statically
but the distribution does not ship static libraries, cvc5 builds GMP
automatically when ``--auto-download`` is given.

SymFPU (Support for the Theory of Floating Point Numbers)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`SymFPU <https://github.com/martin-cs/symfpu/tree/CVC4>`_ is an implementation
of SMT-LIB/IEEE-754 floating-point operations in terms of bit-vector operations.
It is required for supporting the theory of floating-point numbers and can be
downloaded and built automatically.

Optional Dependencies
---------------------

CaDiCaL (Optional SAT solver)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`CaDiCaL <https://github.com/arminbiere/cadical>`_ is a SAT solver that can be

> used for the bit-vector solver. It can be downloaded and built automatically.

Reviewer: Not optional anymore, please move to above section.

    ctest -L regress0                     # run all regression tests in level 0
    ctest -L regress[0-1]                 # run all regression tests in level 0 and 1
    ctest -R regress                      # run all regression tests
    ctest -R regress0                     # run all regression tests in level 0
    ctest -R regress[0-1]                 # run all regression tests in level 0 and 1
    ctest -R regress0/bug288b             # run all tests that match '*regress0/bug288b*'
                                          # > runs regress0/bug288b

Custom Targets
^^^^^^^^^^^^^^

All custom test targets build and run a preconfigured set of tests.

- ``make check [-jN] [ARGS=-jN]``
  The default build-and-test target for cvc5, builds and runs all examples,
  all system and unit tests, and regression tests from levels 0 to 2.

- ``make systemtests [-jN] [ARGS=-jN]``
  Build and run all system tests.

- ``make units [-jN] [ARGS=-jN]``
  Build and run all unit tests.

- ``make regress [-jN] [ARGS=-jN]``
  Build and run regression tests from levels 0 to 2.

- ``make runexamples [-jN] [ARGS=-jN]``
  Build and run all examples.

> - ``make coverage [-jN] [ARGS=-jN]``

Reviewer: ```suggestion
- ``make coverage-test [-jN] [ARGS=-jN]``
```

https://github.com/CVC4/homebrew-cvc4 .
To build a static binary for macOS, use:
``./configure.sh --static --no-static-binary``.

Cross-compiling for Windows
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Cross-compiling cvc5 with Mingw-w64 can be done as follows:

.. code:: bash

  ./configure.sh --win64 --static <configure options...>

  cd <build_dir>   # default is ./build
  make             # use -jN for parallel build with N threads

The built binary ``cvc5.exe`` is located in ``<build_dir>/bin`` and the cvc5
library can be found in ``<build_dir>/lib``.

Build dependencies
------------------

cvc5 makes uses of a number of tools and libraries. Some of these are required
while others are only used with certain configuration options. If
``--auto-download`` is given, cvc5 can automatically download and build most
libraries that are not already installed on your system. If your libraries are
installed in a non-standard location, you can use ``--dep-path`` to define an
additional search path for all dependencies. Versions given are minimum
versions; more recent versions should be compatible.

- `GNU C and C++ (gcc and g++) <https://gcc.gnu.org>`_

>   or `Clang <https://clang.llvm.org>`_ (reasonably recent versions)

Reviewer 2: Instead of "reasonably recent versions", could we say that we require "GCC >= 7 and Clang >= 5"?

- `CMake >= 3.9 <https://cmake.org>`_

> - `Python 3.x <https://www.python.org>`_

Reviewer 2: As discussed in #devs, let's require Python 3.6.

used to provide command editing, tab completion, and history functionality at
the cvc5 prompt (when running in interactive mode).  Check your distribution for
a package named "libedit-dev" or "libedit-devel" or similar.

Google Test Unit Testing Framework (Unit Tests)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`Google Test <https://github.com/google/googletest>`_ is required to optionally
run cvc5's unit tests (included with the distribution). See :ref:`Testing cvc5
<testing-cvc5>` below for more details.

Language bindings
-----------------

cvc5 provides a complete and flexible C++ API (see ``examples/api`` for
examples). It further provides Java (see ``examples/SimpleVC.java`` and
``examples/api/java``) and Python (see ``examples/api/python``) API bindings.

Configure cvc5 with ``configure.sh --<lang>-bindings`` to build with language
bindings for ``<lang>``.

Dependencies for Language Bindings
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- Python

  - `Cython <https://cython.org/>`_
  - `scikit-build <https://pypi.org/project/scikit-build/>`_
  - `pytest <https://docs.pytest.org/en/6.2.x/>`_

If you're interested in helping to develop, maintain, and test a language

> binding, please contact one of the project leaders via `our issue tracker

Reviewer 3: I would get rid of the `project leaders`:
```suggestion
binding, please contact the cvc5 team via `our issue tracker
```

<https://github.com/cvc5/cvc5/issues>`_.

Building the API Documentation
------------------------------

Building the API documentation of cvc5 requires the following dependencies:

- `Doxygen <https://www.doxygen.nl>`_
- `Sphinx <https://www.sphinx-doc.org>`_,
  `sphinx-tabs <https://sphinx-tabs.readthedocs.io/>`_,
  `sphinxcontrib-bibtex <https://sphinxcontrib-bibtex.readthedocs.io>`_
- `Breathe <https://breathe.readthedocs.io>`_

To build the documentation, configure cvc5 with ``./configure.sh --docs``.
Building cvc5 will then include building the API documentation.

The API documentation can then be found at
``<build_dir>/docs/sphinx/index.html``.

> To only build the documentation, change to the build directory and call ``make
> docs``.

Reviewer 3: ```suggestion
To only build the documentation, change to the build directory and call
``make docs``.
```

To build the documentation for GitHub pages, change to the build directory and
call ``make docs-gh``. The content of directory ``<build_dir>/docs/sphinx-gh``
can then be copied over to GitHub pages.

Building the Examples
---------------------

See ``examples/README.md`` for instructions on how to build and run the
examples.

.. _testing-cvc5:

Testing cvc5
------------

> We use ``ctest`` as test infrastructure, for all command-line options of ctest,

Reviewer 3: ```suggestion
We use ``ctest`` as test infrastructure. For all command-line options of ctest,
```

Building cvc5
-------------

.. code:: bash

    ./configure.sh
        # use --prefix to specify an install prefix (default: /usr/local)
        # use --name=<PATH> for custom build directory
        # use --auto-download to download and build missing, required or
        #   enabled, dependencies
    cd <build_dir>   # default is ./build
    make             # use -jN for parallel build with N threads
    make check       # to run default set of tests
    make install     # to install into the prefix specified above

All binaries are built into ``<build_dir>/bin``, the cvc5 library is built into
``<build_dir>/lib``.

Supported Operating Systems
---------------------------

cvc5 can be built natively on Linux and macOS, cross-compilation is possible for

> Windows using Mingw-w64. cvc5 also supports cross-compilation for Arm64 systems.

Reviewer 2: ```suggestion
Windows using Mingw-w64. cvc5 also supports cross-compilation for ARM64 systems.
```

- `SymFPU <https://github.com/martin-cs/symfpu/tree/CVC4>`_

ANTLR 3.4 parser generator
^^^^^^^^^^^^^^^^^^^^^^^^^^

For most systems, the package manager no longer contains pre-packaged versions
of ANTLR 3.4. With ``--auto-download``, cvc5 will automatically download and
build ANTLR 3.4.

GMP (GNU Multi-Precision arithmetic library)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

GMP is usually available on your distribution and should be used from there. If
it does not, or you want to cross-compile, or you want to build cvc5 statically
but the distribution does not ship static libraries, cvc5 builds GMP
automatically when ``--auto-download`` is given.

SymFPU (Support for the Theory of Floating Point Numbers)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`SymFPU <https://github.com/martin-cs/symfpu/tree/CVC4>`_ is an implementation
of SMT-LIB/IEEE-754 floating-point operations in terms of bit-vector operations.
It is required for supporting the theory of floating-point numbers and can be
downloaded and built automatically.

Optional Dependencies
---------------------

CaDiCaL (Optional SAT solver)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> `CaDiCaL <https://github.com/arminbiere/cadical>`_ is a SAT solver that can be
> used for solving non-incremental bit-vector problems with eager bit-blasting.
> This dependency may improve performance. It can be downloaded and built
> automatically. Configure cvc5 with ``configure.sh --cadical`` to build with this
> dependency.

Reviewer 2: ```suggestion
`CaDiCaL <https://github.com/arminbiere/cadical>`_ is a SAT solver that can be
is used for the bit-vector solver. It can be downloaded and built
automatically.
```

Reviewer 2: Also, it should be moved up to the required dependencies.

^^^^^^^^^^^^^^^^^^^^^^^^^^

For most systems, the package manager no longer contains pre-packaged versions
of ANTLR 3.4. With ``--auto-download``, cvc5 will automatically download and
build ANTLR 3.4.

GMP (GNU Multi-Precision arithmetic library)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

GMP is usually available on your distribution and should be used from there. If

> it does not, or you want to cross-compile, or you want to build cvc5 statically

Reviewer 2: ```suggestion
it is not, or you want to cross-compile, or you want to build cvc5 statically
```
